<a href="https://colab.research.google.com/github/Aafreen2603/Handwritten-digit-recognition-using-CNN/blob/main/Handwritten_digit_recognition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Handwritten Digit Recognition 
### Deep Learning Project
Libraries, Tools used:
*   Python programming
*   Deep learning (Convolutional Neural Networks) with Keras library 
*   Tensorflow and Keras libraries
*   the Tkinter library for building GUI




To implement a **handwritten digit recognition app** which uses the image of a digit and recognizes the digit present in the image.

This will be done using **Convolutional Neural Networks** which is a type of deep neural network.



1. Import the libraries and load the dataset into training and test data
2. Preprocess the data
3. Create the model
4. Train the model using training data
5. Evaluate the model using testing data


### Dataset used: The MNIST dataset

The MNIST dataset contains 60,000 training images of handwritten digits from zero to nine and 10,000 images for testing. So, the MNIST dataset has 10 different classes. The handwritten digits images are represented as a 28×28 matrix where each cell contains grayscale pixel value.

## 1. Importing the libraries and loading dataset
First, we are going to import all the modules that we are going to need for training our model. The Keras library already contains the MNIST dataset. So we can import the dataset and start working with it. 

The **mnist.load_data()** method returns us the training data, its labels and also the testing data and its labels.




In [ ]:
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

from keras.utils import np_utils

In [ ]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
print(x_train.shape, y_train.shape)

11501568/11490434 [==============================] - 0s 0us/step
(60000, 28, 28) (60000,)


## 2. Data Preprocessing
The image data cannot be fed directly into the model so we need to perform some operations and process the data to make it ready for our neural network. 
The dimension of the training data is (60000,28,28). 
The CNN model will require one more dimension so we reshape the matrix to shape (60000,28,28,1).

In [ ]:
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)
input_shape = (28, 28, 1)

In [ ]:
# convert class vectors to binary class matrices
y_train = keras.utils.np_utils.to_categorical(y_train, num_classes=None)
y_test = keras.utils.np_utils.to_categorical(y_test,  num_classes=None)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


## 3. Create the model
A CNN model generally consists of convolutional and pooling layers. It works better for data that are represented as grid structures, this is the reason why CNN works well for image classification problems. The dropout layer is used to deactivate some of the neurons and while training, it reduces offer fitting of the model. 
We will then compile the model with the **Adadelta optimizer.**

In [ ]:
batch_size = 128
num_classes = 10
epochs = 10

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),activation='relu',input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

#model.compile(loss=keras.losses.categorical_crossentropy,optimizer=keras.optimizers.Adadelta(),metrics=['accuracy'])
model.compile(loss='categorical_crossentropy', optimizer='adadelta',metrics=['accuracy'])

## 4. Train the model with training set
The model.fit() function of Keras will start the training of the model. It takes the training data, validation data, epochs, and batch size.

It takes some time to train the model. After training, we save the weights and model definition in the ‘mnist.h5’ file.

In [ ]:
hist = model.fit(x_train, y_train,batch_size=batch_size,epochs=epochs,verbose=1,validation_data=(x_test, y_test))
print("The model has successfully trained")

model.save('mnist.h5')
print("Saving the model as mnist.h5")

Epoch 1/10
469/469 [==============================] - 171s 361ms/step - loss: 2.2670 - accuracy: 0.1616 - val_loss: 2.2040 - val_accuracy: 0.4045
Epoch 2/10
469/469 [==============================] - 163s 347ms/step - loss: 2.1648 - accuracy: 0.3230 - val_loss: 2.0785 - val_accuracy: 0.5917
Epoch 3/10
469/469 [==============================] - 162s 345ms/step - loss: 2.0332 - accuracy: 0.4525 - val_loss: 1.9103 - val_accuracy: 0.6761
Epoch 4/10
469/469 [==============================] - 160s 341ms/step - loss: 1.8559 - accuracy: 0.5514 - val_loss: 1.6883 - val_accuracy: 0.7297
Epoch 5/10
469/469 [==============================] - 160s 341ms/step - loss: 1.6417 - accuracy: 0.6136 - val_loss: 1.4299 - val_accuracy: 0.7727
Epoch 6/10
469/469 [==============================] - 158s 338ms/step - loss: 1.4170 - accuracy: 0.6579 - val_loss: 1.1792 - val_accuracy: 0.7975
Epoch 7/10
469/469 [==============================] - 160s 341ms/step - loss: 1.2153 - accuracy: 0.6873 - val_loss: 0.9755 -

We can see that after each epoch, the accuracy increases and the loss decreases.

## 5. Model Evaluation using the testing set

We have 10,000 images in our testing dataset which will be used to evaluate how good our model works. 


In [ ]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.6459946036338806
Test accuracy: 0.8478999733924866


Our model has an accuracy of 84.79% on the test set. To increase this we can increase the number of epochs.